architecture-decisions/2024-06-06_arc4-by-default.md

```diff
@@ -0,0 +1,153 @@
+
+The term ARC4 and ABI are sometimes used interchangeably however ABI refers to the generic concept of an Application Binary Interface (what ARC4 provides) and ARC4 being the specific implementation we have of an ABI on Algorand at this moment. 
+
+
+## Requirements
+
+- It must be possible to define a contract with ARC4 routed methods
+- It must be possible to define a contract which doesn't use ARC4 routing
+
+## Principles
+
+- **[AlgoKit Guiding Principles](https://github.com/algorandfoundation/algokit-cli/blob/main/docs/algokit.md#guiding-principles)** - specifically Seamless onramp, Leverage existing ecosystem, Meet devs where they are
+- **[Algorand Python Principles](https://algorandfoundation.github.io/puya/principles.html#principles)**
+- **[Algorand TypeScript Guiding Principles](../README.md#guiding-principals)**
+
+## Options
+
+### Option 1 - ARC4 by default
+
+The approach taken by TealScript is that contracts are ARC4 by default. All public methods are exposed as ABI methods (callable from outside the contract) and one must explicitly decorate a method that should not be ABI routed. Only one non abi routed method can be defined per on completion action and this method is invoked only if no other ABI routed methods handle the call first. It is not necessary to decorate an ABI method unless an on completion action that is not `NoOp` is required. 
+
+```ts
+class NonABIExample extends Contract {
+   private add(x: uint64, y: uint64): uint64 {
+      return x + y;
+   }
+
+   abiAdd(x: uint64, y: uint64): uint64 {
+      return this.add(x, y);
+   }
+
+   @nonABIRouterFallback.call('NoOp')
+   nonAbiAdd(): void {
+      const x = btoi(this.txn.applicationArgs![0]);
+      const y = btoi(this.txn.applicationArgs![1]);
+      log(itob(this.add(x, y)));
+   }
+}
+```
+
+Handling of 'other' on completion actions can be done via decorators however the prescribed way in the documentation is to implement a [well-known](https://tealscript.netlify.app/guides/lifecycle/) method on your contract.
+
+Pros:
+ - A developer new to Algorand does not need to learn much about ARC4 in order to implement an ARC4 compliant contract.
+ - Flexibility exists for advanced devs to 'break out' of the ARC4 default when desired (kind of)
+
+Cons:
+ - The current implementation in TealScript outputs a lot of redundant teal in scenarios where an advanced user wants to do their own routing
+ - Doesn't give us a low friction path forward if ARC4 were to be replaced, or significantly updated
```

Comment:
I think the chance of ARC4 getting replaced or significantly updated is close to zero. There's a graveyard of ARCs attempting to do so as evidence lol. Also, ARC4 is such a low level detail that has such strong effects up the entire chain of infrastructure that a change to ARC4 is essentially a breaking change to everything. Thus if we want to support a new ABI, having a backwards incompatible way of doing that isn't unreasonable in my opinion. 

To be clear, there's a difference between changing the existing ABI and building on top of it. Maybe we have a new type `timestamp` that is used as a timestamp, but in the ARC4 signature it is still `uint64` and encoded as such. `timestamp` would be reflected at higher levels like the language and ARC56. These sort of changes are possible without having to break anything.